### PR TITLE
Pass withOnyx instance props to selectors

### PR DESCRIPTION
**Pull request by tgolen**
### Details
This is needed for threads. See the discussion [here](https://expensify.slack.com/archives/C03TQ48KC/p1682963933450719?thread_ts=1682959473.757919&cid=C03TQ48KC)

### Fixes
$ https://github.com/Expensify/App/issues/18489

### Automated Tests
Everything is covered by tests

### Linked PRs
@grgia will be implementing this

### CODE REVIEW

**Review by grgia — APPROVED** [review]

File: lib/Onyx.js [hunk 1]
```diff
@@ -662,9 +664,10 @@ function getCollectionDataAndSendAsObject(matchingKeys, mapping) {
  *  component
  * @param {Boolean} [mapping.waitForCollectionCallback] If set to true, it will return the entire collection to the callback as a single object
  * @param {String|Function} [mapping.selector] THIS PARAM IS ONLY USED WITH withOnyx(). If included, this will be used to subscribe to a subset of an Onyx key's data.
- *       If the selector is a string, the selector is passed to lodashGet on the sourceData. If the selector is a function, the sourceData is passed to the selector and should return the
- *       simplified data. Using this setting on `withOnyx` can have very positive performance benefits because the component will only re-render when the subset of data changes.
- *       Otherwise, any change of data on any property would normally cause the component to re-render (and that can be expensive from a performance standpoint).
+ *       If the selector is a string, the selector is passed to lodashGet on the sourceData. If the selector is a function, the sourceData and withOnyx instance props are
```
Comment (PauloGasparSv) [inline lib/Onyx.js:667]: Hey @tgolen this comment/description also exists inside [API.md](https://github.com/Expensify/react-native-onyx/blob/58bc7cc22947f87f594edfedd8f0fbcccec8b30c/API.md?plain=1&#L131). Should we update this description there too?

**Review by PauloGasparSv — DISMISSED** [review]
LGTM, left a [single comment](https://github.com/Expensify/react-native-onyx/pull/247/files#r1186211659) but NAB IMO

File: lib/Onyx.js [hunk 2]
```diff
@@ -62,11 +63,12 @@ const getSubsetOfData = (sourceData, selector) => (_.isFunction(selector)
  * The resulting collection will only contain items that are returned by the selector.
  * @param {Object} collection
  * @param {String|Function} selector (see method docs for getSubsetOfData() for full details)
+ * @param {Object} [withOnyxInstanceProps]
```
Comment (PauloGasparSv) [inline lib/Onyx.js:66]: Same as [this comment](https://github.com/Expensify/react-native-onyx/pull/247/files#r1186211659) should we update API.md to include that new param?

File: lib/Onyx.js [hunk 3]
```diff
@@ -48,12 +48,13 @@ const deferredInitTask = createDeferredTask();
  * Uses a selector string or function to return a simplified version of sourceData
  * @param {Mixed} sourceData
  * @param {String|Function} selector
+ * @param {Object} [withOnyxInstanceProps]
```
Comment (PauloGasparSv) [inline lib/Onyx.js:51]: Same as [this comment](https://github.com/Expensify/react-native-onyx/pull/247/files#r1186211659) should we update API.md to include that new param?

**Review by PauloGasparSv — DISMISSED** [review]
LGTM!

**Comment by tgolen** [comment]
OK, I found with @grgia that we needed to pass the state instead of props, and I also found a spot where I wasn't passing the parameter to the selector properly. I've updated tests and docs

**Review by PauloGasparSv — APPROVED** [review]

**Review by chiragsalian — APPROVED** [review]
havent tested, LGTM. this will be a nice addition.

**Comment by JmillsExpensify** [comment]
What's left to merge this one?

**Comment by tgolen** [comment]
@grgia has been testing this pretty extensively on a local branch. @grgia Do you think it's working as intended or does this need more changes before it can be merged?

**Comment by grgia** [comment]
@tgolen I wasn't able to get it working on my branch using the selector / props - HOC is here if you want to take a look - https://github.com/Expensify/App/pull/18638

**Comment by grgia** [comment]
I'm not sure it's working as expected, I only seem to get `{isLoading: true}` no matter the key/number of onyx connect/prop combo

**Comment by tgolen** [comment]
One way I tested this to ensure it was working was to edit `HeaderView.js` and:

1. Remove all keys from `withOnyx()` config
2. Added this single key in the `withOnyx()` config

```
        reports: {
            key: ONYXKEYS.COLLECTION.REPORT,
            selector: (reports, props) => {
                console.log('PROPS', props);
                return props && props.report && _.findWhere(reports, {reportID: props.report.reportID})
            }
        }
```

3. Verified that I saw it was able to find the report and `{loading: false}` eventually appeared in the console

**Comment by tgolen** [comment]
Since this is working with the unit tests and in the real-world, I'm going to go ahead and merge this. Maybe one day I will get the chance to implement it in NewDot

**Pull request merged**